core/scripts/Voting.js

```diff
@@ -91,104 +90,117 @@ class EmailSender {
-    const topicHash = computeTopicHash(request, roundId);
-    await this.encryptedSender.sendMessage(this.account, topicHash, encryptedMessage, { from: this.account });
-  }
+    for (let i = 0; i < requests.length; i++) {
+      // Skip commits if a message already exists for this request.
+      // This does not check the existence of an actual commit.
+      if (await this.getMessage(requests[i], roundId)) {
+        continue;
+      }
 
-  async runCommit(request, roundId) {
-    const persistedVote = await this.readVote(request, roundId);
-    // If the vote has already been persisted and committed, we don't need to recommit.
-    // TODO: we may want to add a feature in the future where we ensure that the persisted vote matches the commit, and
-    // if it does not, we commit the persisted vote.
-    if (persistedVote) {
-      return false;
+      commitments.push(await this.constructCommitment(requests[i], roundId));
     }
-    const fetchedPrice = await fetchPrice(request);
-    const salt = web3.utils.toBN(web3.utils.randomHex(32));
 
-    const hash = web3.utils.soliditySha3(fetchedPrice, salt);
-    await this.voting.commitVote(request.identifier, request.time, hash, { from: this.account });
+    // Always call `batchCommit`, even if there's only one commitment. Difference in gas cost is negligible.
+    await this.voting.batchCommit(commitments, { from: this.account });
 
-    // Persist the vote only after the commit hash has been sent to the Voting contract.
-    await this.writeVote(request, roundId, { price: fetchedPrice.toString(), salt: salt.toString() });
-    return true;
+    return commitments.length;
   }
 
-  async runReveal(request, roundId) {
-    const persistedVote = await this.readVote(request, roundId);
-    // If no vote was persisted and committed, then we can't reveal.
-    if (!persistedVote) {
-      return false;
+  async constructReveal(request, roundId) {
+    const encryptedCommit = await this.getMessage(request, roundId);
+
+    let vote;
+
+    // Catch messages that are indecipherable and handle by skipping over the request.
+    try {
+      const { privateKey } = await deriveKeyPairFromSignature(web3, getKeyGenMessage(roundId), this.account);
+      vote = JSON.parse(await decryptMessage(privateKey, encryptedCommit));
+    } catch (e) {
+      console.error("Failed to decrypt message:", encryptedCommit, "\n", e);
+      return null;
     }
-    const hasRevealedVote = await this.voting.hasRevealedVote(request.identifier, request.time, { from: this.account });
-    // If we've already revealed, no need to re-reveal.
-    if (hasRevealedVote) {
-      return false;
+
+    return {
+      identifier: request.identifier,
+      time: request.time,
+      price: vote.price.toString(),
+      salt: web3.utils.hexToNumberString(vote.salt)
+    };
+  }
+
+  async runBatchReveal(requests, roundId) {
+    const reveals = [];
+
+    let reveal;
+    for (let i = 0; i < requests.length; i++) {
```

Comment:
nit: would a `for ... of` loop here make sense?

Reply:
haha, yeah, I keep on forgetting that this exists.